src/Compiler/Checking/PostInferenceChecks.fs

```diff
@@ -532,8 +532,16 @@ let CheckTypeForAccess (cenv: cenv) env objName valAcc m ty =
             | ValueSome tcref ->
                 let thisCompPath = compPathOfCcu cenv.viewCcu
                 let tyconAcc = tcref.Accessibility |> AccessInternalsVisibleToAsInternal thisCompPath cenv.internalsVisibleToPaths
-                if isLessAccessible tyconAcc valAcc then
-                    errorR(Error(FSComp.SR.chkTypeLessAccessibleThanType(tcref.DisplayName, (objName())), m))
+                
+                // Skip accessibility checks for compiler-generated pattern inputs, which all have names
+                // starting with "val patternInput". This allows pattern matching to work properly with private
+                // types when the pattern is used within the same module scope.
+                let isCompilerGeneratedPatternInput = 
+                    let name: string = objName()
+                    name.StartsWith("val patternInput")
```

Comment:
Don't we still have the `ValRef` itself in scope where this function (`CheckTypeForAccess`) is called in `CheckBinding`? Could we just look at `vref.IsCompilerGenerated` there instead?

Reply:
Yeah it seems that we can use `vref.IsCompilerGenerated` for the CheckBinding case.